### PR TITLE
added canister-based-wallet-architects.md and diagram pic

### PR DESCRIPTION
**Pull request by neeboo**
canister based wallet architects and diagram

### CODE REVIEW

**Review by frederikrothenberger — COMMENTED** [review]
Thanks a lot @neeboo! This is an awesome start to the solution discussion. I really like it. 😄👍

In particular, I think it's really elegant that this solution checks both those boxes:
- [x] no changes to the replica
- [x] centralized permission checking (i.e. the target is _not_ responsible for checking authorization)

I would suggest the following changes to this PR:
* Add a "Proposed Solutions" section to the end of https://github.com/dfinity/wg-identity-authentication/blob/main/topics/interoperability.md and link your document there.
* Add a section "Next Steps / Open Questions" to this documents so we can discuss the next steps to make further progress.
  * In particular I think we should start defining the interfaces that should be standardized. I think the following interactions need to be specified by this solution:
    * Interactions between IDP Frontend <-> Frontend to request permission / capabilities
    * Interactions between IDP Frontend <-> Target to fetch available / valid permissions / capabilities
    * Interactions between IDP Frontend <-> Wallet canister to manage permissions, see log of past authorizations, etc.
  * Discuss concrete usage scenarios using this architecture and provide end-to-end examples.
* This PR talks a lot about NNS / II / Me etc. I think it would be better to use more neutral wording and just use identity provider (IDP) in those cases.
* In the diagram (or a separate one) it would be nice to show that many front-ends will connect to the wallet canister this way.

And maybe just a general question:
Do you expect the IDP front end to spawn the wallet canister or should it be that each use can also bring their own wallet as long as it implements the required interfaces? I think this would be an awesome property to have to give the most control possible to the users (i.e. build your own if you don't trust the default wallet implementation).

**Comment by neeboo** [comment]
> And maybe just a general question: Do you expect the IDP front end to spawn the wallet canister or should it be that each use can also bring their own wallet as long as it implements the required interfaces? I think this would be an awesome for giving full control to the users (i.e. build your own if you don't trust the default wallet implementation).

I would choose the latter if possible, however maybe we need some registry to query wallet list from his/her controller principal id. But maybe we can allow users to input and save their wallet canister id with owner check.

File: topics/canister-based-wallet-architects.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+   - **Expiration**, for how long the authorization and DIP will live.
+   - **Target canisters**, which canisters are allowed to call, with ICScan links.
+   - **Target functions(capabilities)**, a list of function names and usage, the user can choose to call or not
+   - Other necessary information for example trusted/audited badge.
+   - Other validation processes can be run here, for example, a simulation call by a trusted 3rd party.
+4. If the user confirms, the DIP can be added to the wallet canister, and all valid authorization calls will be saved before expiration. And for fine control, we can also provide a UI that allows users to change authorized functions at his/her will.
+5. If the user rejects it, the `connect wallet` process will be terminated.
+6. When the wallet call happens, a few arguments will be passed to function, including:
+
+### Limitations and potential problems
```
[the remark this answers is not in the document]
Comment (neeboo) [reply topics/canister-based-wallet-architects.md:78]: Will add these to next commit

File: topics/canister-based-wallet-architects.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+Now we can design the authentication and authorize flow:
+
+1. User logins with II, the DApp frontend will receive a delegation identity principal (DIP). For DApps don't need to use the wallet, the DIP will work as the normal function caller.
+2. When DApp wants to access the user's wallet canister, there is another button we call `Connect Wallet`. When this button is clicked, a new window appears and shows the wallet canister owned by the NNS delegation identity.
+3. Also in this window, there are a few necessary pieces of information will be shown to the user:
+   - **DIP**, whether it can be added to the wallet canister's ACL.
+   - **Expiration**, for how long the authorization and DIP will live.
+   - **Target canisters**, which canisters are allowed to call, with ICScan links.
+   - **Target functions(capabilities)**, a list of function names and usage, the user can choose to call or not
```
[the remark this answers is not in the document]
Comment (neeboo) [reply topics/canister-based-wallet-architects.md:71]: Yes I think it's important to get human readable information to the IDP frontend. Maybe we can follow some standard already existed / used in industry without create something new.  @domwoe had done some research on SIWE and CACAOS standard. Might be helpful we refer this repo here: https://github.com/domwoe/siwe_authz.

File: topics/canister-based-wallet-architects.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+1. The first problem is issued in https://github.com/dfinity/wg-identity-authentication/issues/5, generally, it's protocol level requirement and solution for global delegation and interoperability of user identity. We believe DFINITY will try its best to make this happens.
+
+2. The second problem is related to the first one, but it's the interaction pattern of the canister. [NNS wallet](https://nns.ic0.app) and [AstroX ME Wallet](https://app.astrox.me) use` principal<>delegation identity<>ledger` to create wallet instances for users. And developers try to access users' NNS or ME wallet from their DApp, and they want to use wallet identity to call their DApp backend canister because they assume that assets are stored in NNS or ME wallet. Because both NNS and ME uses session key in their frontend, and for security reason (sessionKey exposure), NNS and ME wallet can not provide such kind of function call.
```
Comment (marydwyer) [inline topics/canister-based-wallet-architects.md:20]: How would you reframe this problem as a requirement? I'd like to add it to the [list of interoperability requirements](https://github.com/dfinity/wg-identity-authentication/blob/main/topics/interoperability.md#interoperability-of-services-using-user-identity).

File: topics/canister-based-wallet-architects.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+3. One of the authorized function calls is `wallet_call`, here we use `async/await` to call cdk API, which is `ic_cdk::api::call::call128`. [Example](https://github.com/dfinity/cycles-wallet/blob/fa86dd3a65b2509ca1e0c2bb9d7d4c5be95de378/wallet/src/lib.rs#L898)
+
+The trusted setting here is to allow a user to `own` this canister because we assume this canister works as a private wallet, which should be served privately to the user him/herself, not to the public. Only the user can have the ability to create/upgrade/stop/delete this canister, the power can be delegated to a trusted 3rd party service under the user's permission. We shall have another topic to discuss this scheme.
+
+We have 2 assumptions here:
+
+- We assume the controller of this wallet canister is the user's NNS delegation Identity.
+- The whole flow is running on the user's browser
```
Comment (marydwyer) [inline topics/canister-based-wallet-architects.md:61]: What are the risks and benefits of running the whole flow on the browser?

**Comment by marydwyer** [comment]
Thanks for the great proposal @neeboo! I'm looking forward to continuing to collaborate on it. A couple of notes:

-  I'm having a hard time following the diagram. You illustrate the elements of the wallet canister very well. However, I'd like to see the user, dapp A, dapp B, and identity provider and understand when the user is using their principal for dapp A, dapp B or their delegation identity principal (DIP). I am not following the relationship between the wallet canister, targets, and frontend. 

- Does this solution depend on the development of any other technologies? If so, could we add a dependencies section?

**Comment by neeboo** [comment]
> Thanks for the great proposal @neeboo! I'm looking forward to continuing to collaborate on it. A couple of notes:
> 
> * I'm having a hard time following the diagram. You illustrate the elements of the wallet canister very well. However, I'd like to see the user, dapp A, dapp B, and identity provider and understand when the user is using their principal for dapp A, dapp B or their delegation identity principal (DIP). I am not following the relationship between the wallet canister, targets, and frontend.
> * Does this solution depend on the development of any other technologies? If so, could we add a dependencies section?

I have a little demo showing here: https://github.com/AstroxNetwork/wallet_call

**Comment by frederikrothenberger** [comment]
Given this PR has not made progress in over a year, I think it is appropriate to close it. Feel free to reopen if you disagree.

**Pull request closed without merging**